runtime/lua/vim/lsp.lua

```diff
@@ -827,23 +833,12 @@ do
     util.buf_versions[bufnr] = changedtick
     -- Lazy initialize these because clients may not even need them.
     local incremental_changes = once(function(client)
-      local size_index = encoding_index[client.offset_encoding]
-      local length = select(size_index, old_byte_size, old_utf16_size, old_utf32_size)
-      local lines = nvim_buf_get_lines(bufnr, firstline, new_lastline, true)
-
-      -- This is necessary because we are specifying the full line including the
-      -- newline in range. Therefore, we must replace the newline as well.
-      if #lines > 0 then
-       table.insert(lines, '')
-      end
-      return {
-        range = {
-          start = { line = firstline, character = 0 };
-          ["end"] = { line = lastline, character = 0 };
-        };
-        rangeLength = length;
-        text = table.concat(lines, '\n');
-      };
+      local lines = nvim_buf_get_lines(bufnr, 0, -1, true)
+      local startline =  math.min(firstline + 1, math.min(#client._cached_buffers[bufnr], #lines))
+      local endline =  math.min(-(#lines - new_lastline), 0)
+      local incremental_change = vim.lsp.util.compute_diff(client._cached_buffers[bufnr], lines, startline, endline)
```

Comment:
I've had a `nil` error somewhere around here today. It was one of the `#` calls. Just as a heads up. I'll try to reproduce and provide more information later.


Another thought I had: Should we clear the cached buffers on detach?

Reply:
Oh, I just changed the end line  calculation so let me know if it pops up again

Reply:
Oh the bug is because on new buffer attaches the last buffer cache table isn't being created

Reply:
@mfussenegger I think I fixed it, and I added the clearing of the cache on detach

Reply:
> Oh the bug is because on new buffer attaches the last buffer cache table isn't being created

Oh yes, looks like that was the problem.
I think it's fixed.